core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -109,130 +109,175 @@ final class FiberRefs private (
+            if (emptyChildStack) parentFiberRefs
+            else
+              parentFiberRefs.updated(
+                childFiberRef,
+                FiberRefStack.init(fiberId, childFiberRef.join(childFiberInitialValue, childStackCurrentValue))
+              )
+          case parentStack: FiberRefStack[Any] =>
+            val ancestor =
+              findAncestor(childFiberInitialValue)(
+                parentStack.stack(),
+                parentStack.depth,
+                childStack.stack(),
+                childStack.depth
+              )
+            val patch    = childFiberRef.diff(ancestor, childStackCurrentValue)
+            val oldValue = parentStack.value
+            val newValue = childFiberRef.join(oldValue, childFiberRef.patch(patch)(oldValue))
+
+            if (oldValue == newValue) parentFiberRefs
+            else {
+              val newStack =
+                if (parentStack.fiberId == fiberId) parentStack.updateValue(newValue)
+                else parentStack.put(fiberId, newValue)
+
+              parentFiberRefs.updated(childFiberRef, newStack)
             }
         }
+      }
     }
 
-    if (self.fiberRefLocals eq fiberRefLocals0) self
-    else FiberRefs(fiberRefLocals0)
+    if (self.fiberRefLocals eq newFiberRefLocals) self
+    else FiberRefs(newFiberRefLocals)
   }
 
   def setAll(implicit trace: Trace): UIO[Unit] =
     ZIO.foreachDiscard(fiberRefs) { fiberRef =>
       fiberRef.asInstanceOf[FiberRef[Any]].set(getOrDefault(fiberRef))
     }
 
-  override final def toString(): String = fiberRefLocals.mkString("FiberRefLocals(", ",", ")")
+  override def toString: String = fiberRefLocals.mkString("FiberRefLocals(", ",", ")")
 
   def updatedAs[@specialized(SpecializeInt) A](
     fiberId: FiberId.Runtime
   )(fiberRef: FiberRef[A], value: A): FiberRefs = {
-    val oldEntry = fiberRefLocals.getOrElse(fiberRef, null)
+    val oldStack = fiberRefLocals.getOrElse(fiberRef, null).asInstanceOf[FiberRefStack[A]]
 
-    val newEntry =
-      if (oldEntry eq null) Value(::(StackEntry(fiberId, value, 0), List.empty), 1)
+    val newStack =
+      if (oldStack eq null) FiberRefStack.init(fiberId, value)
       else {
-        val oldStack = oldEntry.stack.asInstanceOf[::[StackEntry[A]]]
-        val oldDepth = oldEntry.depth
-        if (oldStack.head.id == fiberId) {
-          val oldValue = oldStack.head.value
-          if (oldValue == value) oldEntry
-          else
-            Value(
-              ::(StackEntry(fiberId, value, oldStack.head.version + 1), oldStack.tail),
-              oldEntry.depth
-            )
-        } else if (oldStack.head.value == value)
-          oldEntry
-        else
-          Value(::(StackEntry(fiberId, value, 0), oldStack), oldDepth + 1)
+        if (oldStack.fiberId == fiberId) {
+          if (oldStack.value == value) oldStack else oldStack.updateValue(value)
+        } else if (oldStack.value == value) oldStack
```

Comment:
By the way is it just me or should this condition be first? It seems we check the same thing in L199 and L200, so there's not really any reason to check `oldStack.fiberId == fiberId` when `oldStack.value == value`